{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import h5py \n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from tqdm import tqdm\n",
    "from Bio import SeqIO\n",
    "\n",
    "sns.set(style='white')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "make a df: distance vs attribution codon wise (with and without absolute values)\n",
    "'''\n",
    "record_tr = []\n",
    "sequence = []\n",
    "\n",
    "for record in SeqIO.parse(\"/net/lts2gdk0/mnt/scratch/lts2/nallapar/rb-prof/data/Jan_2024/Lina/reference/ensembl.cds.fa\", \"fasta\"):\n",
    "    record_tr.append(str(record.description.split(' ')[0]))\n",
    "    sequence.append(str(record.seq))\n",
    "\n",
    "# make df\n",
    "df = pd.DataFrame({'Transcript': record_tr, 'Sequence': sequence})\n",
    "\n",
    "print(\"Made Transcript-Sequence DF\")\n",
    "\n",
    "# convert sequence into codon triplets list\n",
    "df['Codon'] = df['Sequence'].apply(lambda x: [x[i:i+3] for i in range(0, len(x), 3)])\n",
    "\n",
    "df_test = pd.read_csv('predictions/exp_dat_testLiver.csv')\n",
    "test_tr = list(df_test['transcript'])\n",
    "\n",
    "int_ds = h5py.File('/net/lts2gdk0/mnt/scratch/lts2/nallapar/rb-prof/data/Jan_2024/Lina/full_Test_LIXG_RiboGL.h5', 'r')\n",
    "len_samples = len(int_ds['node_attr'])\n",
    "int_tr = list(int_ds['transcript'])\n",
    "\n",
    "codon = []\n",
    "distance_ASite = []\n",
    "attribution_vals = []\n",
    "attribution_abs = []\n",
    "\n",
    "for i in tqdm(range(len_samples)):\n",
    "    node_attr_sample = int_ds['node_attr'][i]\n",
    "    # num codons is square root of length of node_attr_sample\n",
    "    num_codons = int(len(node_attr_sample)**0.5)\n",
    "    # convert to 2D array of shape (num_codons, num_codons)\n",
    "    node_attr_sample = node_attr_sample.reshape(num_codons, num_codons)\n",
    "    # get codon sequence for this transcript\n",
    "    codon_seq = df[df['Transcript'] == str(int_tr[i].decode('utf-8'))]['Codon'].values[0]\n",
    "\n",
    "    for j in range(num_codons):\n",
    "        sample_attr = node_attr_sample[j] / np.max(np.abs(node_attr_sample[j]))\n",
    "        sample_attr_mod = np.abs(node_attr_sample[j]) / np.max(np.abs(node_attr_sample[j]))\n",
    "        for k in range(num_codons):\n",
    "            attribution_vals.append(sample_attr[k])\n",
    "            attribution_abs.append(sample_attr_mod[k])\n",
    "            distance_ASite.append(k-j)\n",
    "            codon.append(codon_seq[k])\n",
    "\n",
    "data = {'Distance from A Site': distance_ASite, 'Attribution Normal': attribution_vals, 'Attribution Mod': attribution_abs, 'Codon': codon}\n",
    "df_distCodonAttr = pd.DataFrame(data)\n",
    "\n",
    "df_distCodonAttr.to_csv('final_plots/distCodonAttr.csv')\n",
    "\n",
    "print(\"Saved DF with Distance from A Site vs Attribution (Normal and Mod) for each Codon\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Global Attribution Plot\n",
    "'''\n",
    "# make line plot with seaborn\n",
    "# set size of plot\n",
    "plt.figure(figsize=(10, 6))\n",
    "sns.lineplot(x=df_distCodonAttr['Distance from A Site'], y=df_distCodonAttr['Attribution Mod'], ci='sd', color='#eb4d4b')\n",
    "# crop plot to only show relevant data from -500 to 500\n",
    "plt.xlim(-500, 500)\n",
    "# add vertical bar on x=0\n",
    "plt.axvline(x=0, color='black', linestyle='--')\n",
    "# x and y labels\n",
    "plt.xlabel('Offset from A site (codons)', fontsize=15)\n",
    "plt.ylabel('Attribution', fontsize=15)\n",
    "\n",
    "plt.title('Contribution of positions in the global neighborhood of the A site', fontsize=20)\n",
    "plt.savefig('final_plots/distance_vs_attribution_500Global.svg')\n",
    "plt.savefig('final_plots/distance_vs_attribution_500Global.png')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "local attribution plot from -7 to +7\n",
    "'''\n",
    "\n",
    "# box plot with seaborn from -7 to 7 \n",
    "df_local = df_distCodonAttr[(df_distCodonAttr['Distance from A Site'] >= -7) & (df_distCodonAttr['Distance from A Site'] <= 7)]\n",
    "plt.figure(figsize=(10, 6))\n",
    "# barplot for the values from -7 to 7\n",
    "sns.barplot(x=df_local['Distance from A Site'], y=df_local['Attribution Mod'], color='#eb4d4b')\n",
    "plt.ylim(0, 0.55)\n",
    "\n",
    "plt.xlabel('Offset from A site (codons)', fontsize=15)\n",
    "plt.ylabel('Attribution', fontsize=15)\n",
    "# change x ticks to be from -7 to 7\n",
    "plt.xticks(range(0, 15), ['-7', '-6', '-5', '-4', '-3', 'E', 'P', 'A', '1', '2', '3', '4', '5', '6', '7'])\n",
    "# change y ticks to remove the last value, make that blank\n",
    "plt.yticks(np.arange(0, 0.6, 0.1), ['0', '0.1', '0.2', '0.3', '0.4', ''])\n",
    "# plot title\n",
    "plt.title('Contribution of positions in the local neighborhood of the A site', fontsize=20)\n",
    "plt.savefig('final_plots/distance_vs_attribution_barALocal.svg')\n",
    "plt.savefig('final_plots/distance_vs_attribution_barALocal.png')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "make codon wise attribution plot\n",
    "'''\n",
    "\n",
    "# convert codon column to list\n",
    "# df['Codon'] = df['Codon'].apply(lambda x: x[1:-1].split(','))\n",
    "\n",
    "# only take positions -5 to 5\n",
    "df_localCodon = df_distCodonAttr[(df_distCodonAttr['Distance from A Site'] >= -5) & (df_distCodonAttr['Distance from A Site'] <= 5)]\n",
    "\n",
    "# remove codon with len>3\n",
    "df_localCodon = df_localCodon[df_localCodon['Codon'].apply(lambda x: len(x) == 3)]\n",
    "\n",
    "# codon to int dictionary\n",
    "codon_to_id = {\n",
    "    'GCG': 0,\n",
    "    'GCA': 1,\n",
    "    'GCC': 2,\n",
    "    'GCT': 3,\n",
    "    'TGT': 4,\n",
    "    'TGC': 5,\n",
    "    'GAC': 6,\n",
    "    'GAT': 7,\n",
    "    'GAG': 8,\n",
    "    'GAA': 9,\n",
    "    'TTT': 10,\n",
    "    'TTC': 11,\n",
    "    'GGC': 12,\n",
    "    'GGT': 13,\n",
    "    'GGA': 14,\n",
    "    'GGG': 15,\n",
    "    'CAT': 16,\n",
    "    'CAC': 17,\n",
    "    'ATC': 18,\n",
    "    'ATA': 19,\n",
    "    'ATT': 20,\n",
    "    'AAG': 21,\n",
    "    'AAA': 22,\n",
    "    'TTA': 23,\n",
    "    'TTG': 24,\n",
    "    'CTT': 25,\n",
    "    'CTA': 26,\n",
    "    'CTC': 27,\n",
    "    'CTG': 28,\n",
    "    'ATG': 29,\n",
    "    'AAC': 30,\n",
    "    'AAT': 31,\n",
    "    'CCT': 32,\n",
    "    'CCC': 33,\n",
    "    'CCA': 34,\n",
    "    'CCG': 35,\n",
    "    'CAG': 36,\n",
    "    'CAA': 37,\n",
    "    'CGG': 38,\n",
    "    'AGA': 39,\n",
    "    'CGA': 40,\n",
    "    'AGG': 41,\n",
    "    'CGC': 42,\n",
    "    'CGT': 43,\n",
    "    'TCT': 44,\n",
    "    'TCC': 45,\n",
    "    'TCA': 46,\n",
    "    'TCG': 47,\n",
    "    'AGT': 48,\n",
    "    'AGC': 49,\n",
    "    'ACT': 50,\n",
    "    'ACG': 51,\n",
    "    'ACC': 52,\n",
    "    'ACA': 53,\n",
    "    'GTA': 54,\n",
    "    'GTC': 55,\n",
    "    'GTT': 56,\n",
    "    'GTG': 57,\n",
    "    'TGG': 58,\n",
    "    'TAC': 59,\n",
    "    'TAT': 60,\n",
    "    'TAA': 61,\n",
    "    'TGA': 62,\n",
    "    'TAG': 63\n",
    "}\n",
    "\n",
    "id_to_codon = {v:k for k,v in codon_to_id.items()}\n",
    "\n",
    "# convert codon to int\n",
    "codon_list = list(df_localCodon['Codon'])\n",
    "codonintlist = []\n",
    "for codon in codon_list:\n",
    "    codonintlist.append(codon_to_id[codon])\n",
    "df_localCodon['CodonInt'] = codonintlist\n",
    "# drop codon column\n",
    "df_localCodon = df_localCodon.drop('Codon', axis=1)\n",
    "df_localCodon = df_localCodon.drop('Unnamed: 0', axis=1)\n",
    "\n",
    "# make a heatmap with seaborn, with the mean attribution for each codon at each position    \n",
    "df_localCodon = df_localCodon.groupby(['Distance from A Site', 'CodonInt']).mean().reset_index()\n",
    "# make a table 15 x 64 with the mean attribution for each codon at each position\n",
    "df_table = df_localCodon.pivot(index='Distance from A Site', columns='CodonInt', values='Attribution')\n",
    "plt.figure(figsize=(40, 10))\n",
    "# make a heatmap with seaborn using the table\n",
    "# cmap negative with red, positive with blue\n",
    "# print full df_table\n",
    "# save df table\n",
    "df_table.to_csv('df_table.csv')\n",
    "# impute nans to 0\n",
    "df_table = df_table.fillna(0)\n",
    "# invert the y axis\n",
    "df_table = df_table.iloc[::-1]\n",
    "\n",
    "# remove last three rows\n",
    "df_table = df_table.iloc[:,:-3]\n",
    "\n",
    "ax = sns.heatmap(df_table, cmap='coolwarm', center=0, cbar_kws={'label': ''}, square=True, annot_kws={\"size\": 40})\n",
    "ax.figure.axes[-1].yaxis.label.set_size(40)\n",
    "cbar = ax.collections[0].colorbar\n",
    "# here set the labelsize by 20\n",
    "cbar.ax.tick_params(labelsize=25)\n",
    "plt.xlabel('', fontsize=40)\n",
    "plt.ylabel('Offset from A site (codons)', fontsize=40)\n",
    "plt.title('Contribution of codons in different positions around the A site', fontsize=40)\n",
    "\n",
    "# flip y axis\n",
    "# change y axis \n",
    "plt.yticks(range(0, 11), ['5', '4', '3', '2', '1', 'A', 'P', 'E', '-3', '-4', '-5'], fontsize=30, va='top', rotation=0)\n",
    "\n",
    "codon_to_aa = {\n",
    "        'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "        'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "        'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "        'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                \n",
    "        'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "        'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "        'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "        'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "        'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "        'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "        'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "        'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "        'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "        'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "        'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "        'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W',\n",
    "        'NNG':'R', 'NNC':'T', 'NGT':'S', 'NGA':'R',\n",
    "        'NNT':'Y', 'NGC':'S'\n",
    "    }\n",
    "\n",
    "# group codons by amino acid and order the x ticks\n",
    "aa_list = []\n",
    "codon_list = []\n",
    "for i in range(61):\n",
    "    aa_list.append(codon_to_aa[id_to_codon[i]])\n",
    "    codon_list.append(id_to_codon[i])\n",
    "\n",
    "plt.xticks(range(0, 61), codon_list, rotation=90, fontsize=30, ha='left')\n",
    "\n",
    "plt.savefig('final_plots/distance_vs_attribution_heatmapCodon.png')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "riboclette",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
